src/lib/udev/google_nvme_id

```diff
@@ -0,0 +1,245 @@
+#!/bin/bash
+# Copyright 2020 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Used to generate symlinks for PD-NVMe devices using the disk names reported by
+# the metadata server
+
+# Locations of the script's dependencies
+readonly nvme_cli_bin=/usr/sbin/nvme
+
+# Bash regex to parse device paths and controller identification
+readonly NAMESPACE_NUMBER_REGEX="/dev/nvme[[:digit:]]+n([[:digit:]]+).*"
+readonly PARTITION_NUMBER_REGEX="/dev/nvme[[:digit:]]+n[[:digit:]]+p([[:digit:]]+)"
+readonly PD_NVME_REGEX="sn[[:space:]]+:[[:space]]+nvme_card-pd"
+
+# Globals used to generate the symlinks for a PD-NVMe disk.  These are populated
+# by the identify_pd_disk function and exported for consumption by udev rules.
+ID_SERIAL=''
+ID_SERIAL_SHORT=''
+
+#######################################
+# Helper function to log an error message to stderr.
+# Globals:
+#   None
+# Arguments:
+#   String to print as the log message
+# Outputs:
+#   Writes error to STDERR
+#######################################
+function err() {
+  echo "[$(date +'%Y-%m-%dT%H:%M:%S%z')]: $*" >&2
+}
+
+#######################################
+# Retrieves the device name for an NVMe namespace using nvme-cli.
+# Globals:
+#   Uses nvme_cli_bin
+# Arguments:
+#   The path to the nvme namespace (/dev/nvme0n?)
+# Outputs:
+#   The device name parsed from the JSON in the vendor ext of the ns-id command.
+# Returns:
+#   0 if the device name for the namespace could be retrieved, 1 otherwise
+#######################################
+function get_namespace_device_name() {
+  local nvme_json
+  nvme_json="$("$nvme_cli_bin" id-ns -b "$1" | xxd -p -seek 384 | xxd -p -r)"
+  if [[ $? -ne 0 ]]; then
+    return 1
+  fi
+
+  if [[ -z "$nvme_json" ]]; then
+    err "NVMe Vendor Extension disk information not present"
+    return 1
+  fi
+
+  local device_name
+  device_name="$(echo "$nvme_json" | sed -e 's/.*"device_name":"\([a-zA-Z0-9_-]\+\)".*/\1/g')"
```

Comment:
i think this needs careful consideration. JSON is a structured format, and this is a simple string parser, and it is not a safe way to do the transformation you want. have you considered writing this script in another language? is a dependency on jq possible?

Reply:
It could be re-written in python so long as we could guarantee a stable release version across our supported distros (i.e., if everything had a python3 it shouldn't be problematic, save for the call-out to the shell to invoke nvme-cli which has its own set of problems [not a python expert, so I'm not sure how frowned upon external invocations like that via os are]). 

I'm trying to limit the number of external dependencies required and am not familiar with jq.  Open to your advice/suggestions here, though.

Reply:
well, we actually moved a lot of our software away from python bc of the difficulties in guaranteeing consistency, shipping for different versions, etc. 

i will see how widely available jq is or if there is an alternate suggestion. but json has multiple string representations and it's tough to use something like sed safely with it, especially in a future-proof way. here are three string representations of the same json object, which would have different parsed results using the proposed sed command:

```
{"device_name":"my-pd-test-short-name","disk_type":"PERSISTENT"}
```

```
{
  "device_name": "my-pd-test-short-name",
  "disk_type": "PERSISTENT"
}
```

```
{
  "disk_type":"PERSISTENT",
  "device_name":"my-pd-test-short-name"
}
```

Reply:
Makes sense.  FWIW, if jq turns out to be viable, the replacement is pretty straightforward:

echo "$nvme_json" | jq -r .device_name

Reply:
yes, jq also supports hard-to-sed JSON like objects where the key and value are separated by a newline. unfortunately it doesn't seem universally available on supported images. so for that reason, i propose we just try to make this as safe as possible by adding a grep to select the specific line, and allowing for possible whitespace between key and value. suggest amending to one of:
```
device_name="$(echo "$nvme_json" | grep device_name | sed -e 's/.*"device_name":[  \t]*"\([a-zA-Z0-9_-]\+\)".*/\1/')"
```

```
device_name="$(echo "$nvme_json" | grep device_name | cut -d: -f2 | tr -d ' ",')"
```

Reply:
Updated to the first option and verified in a devcluster